fix flacky test in TestObjectInspectorUtils

bjchris32 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Sort the fields in `getDeclaredNonStaticFields` to fix flaky test.


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Avoid flaky test error.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### Is the change a dependency upgrade?
<!--
If yes, please attach a file with output from mvn dependency:tree to validate a complete upgrade of dependency.
If no, write 'No'.
-->
No


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Run the unit test with nondex.

wuh3 commented:
Hey Shao, this PR looks good to me. The only thing I am curious about is that I wonder if you ran Nondex with multiple times using ```-DnondexRuns=xxx```. If that passed, the fix should be perfect. Besides, I guess we both can improve the description here.

bjchris32 commented:
> Hey Shao, this PR looks good to me. The only thing I am curious about is that I wonder if you ran Nondex with multiple times using `-DnondexRuns=xxx`. If that passed, the fix should be perfect. Besides, I guess we both can improve the description here.

Yes, I had tested with multiple runs with NonDex, and it passes. Did you meet any error?
I also added more lines in the "why" section.
Please feel free to add more if you want.

rRajivramachandran commented on serde/src/test/org/apache/hadoop/hive/serde2/objectinspector/TestObjectInspectorUtils.java:
```diff
@@ -76,7 +76,7 @@ public void testObjectInspectorUtils() throws Throwable {
       List<? extends StructField> fields = soi.getAllStructFieldRefs();
       assertEquals(10, fields.size());
-      assertEquals(fields.get(0), soi.getStructFieldRef("aint"));
+      assertEquals(fields.get(0), soi.getStructFieldRef("astring"));
```
In addition to sorting for ensuring proper order of `getDeclaredFields`, why is this change needed

bjchris32 replied:
The assertion should be changed because the first element has been changed after sorting.

rRajivramachandran commented:
It is a well documented PR. Following are some improvements:
1. Please start the PR with the intent such as, this PR fixes the nondeterministic behavior of the test `testName`, followed by explanation of how it was found using Nondex
2. At this line: "it will make assertions on the returned values of soi.getStructFieldData", kindly provide context on what soi is/code pointer and how this method is related to the following explanation regrading ReflectionStructObjectInspector#getStructFieldData.
3. is defined in hive/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/ReflectionStructObjectInspector.java -> Consider using a code pointer instead of describing this
4. Left a comment in code

bjchris32 commented:
> It is a well documented PR. Following are some improvements:
> 
> 1. Please start the PR with the intent such as, this PR fixes the nondeterministic behavior of the test `testName`, followed by explanation of how it was found using Nondex
> 2. At this line: "it will make assertions on the returned values of soi.getStructFieldData", kindly provide context on what soi is/code pointer and how this method is related to the following explanation regrading ReflectionStructObjectInspector#getStructFieldData.
> 3. is defined in hive/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/ReflectionStructObjectInspector.java -> Consider using a code pointer instead of describing this
> 4. Left a comment in code

@rRajivramachandran thanks for your comment.
1. I've added the test name at the beginning.
2. I put the original `soi.getStructFieldData` wrong. It should be `soi.getAllStructFieldRefs`, and that is why you could not find the connection.
3. I just learned how to include the prominent link in my markdown. Thanks for the heads-up.
4. I've replied in the comment.
Thank you!

wuh3 commented:
I also tried multiple Nondex runs, and got no problem with that. Everything looks good to me. I think we can try the real PR now

zzjas commented:
It would be better to list the tests that this PR fixes in addition to the test class.

You can proceed to open a real PR. Once you open a real PR, please mark this tentative PR as `Opened` in your `tentative_pr.csv` file and also raise a PR to IDoFT marking this as `Opened`. Thanks!

FarmersWrap commented:
This PR looks good to me. I tried 500 times on my VM, no failure was raised by your implementation.

NamanJain34 commented:
The description looks good and as it has been tried 500 times on the VM the changes looks good to me.

njain2208 commented:
The description looks good and as it has been tried 500 times on the VM the changes looks good to me.

saurabh-shetty commented:
The PR looks good to me. Some more suggestions will be:
1) What error did you get with NonDex? 
2) What test cases failed?

bjchris32 commented:
I've addressed the comment on my opened PR here: https://github.com/apache/hive/pull/4798
But, the CI failed in the original repo.
I'm still trying to figure that out.
